Mathlib/RingTheory/FractionalIdeal/Norm.lean

```diff
@@ -0,0 +1,145 @@
+/-
+Copyright (c) 2024 Xavier Roblot. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Xavier Roblot
+-/
+import Mathlib.RingTheory.FractionalIdeal.Basic
+import Mathlib.RingTheory.Ideal.Norm
+
+/-!
+
+# Fractional ideal norms
+
+This file defines the absolute ideal norm of a fractional ideal `I : FractionalIdeal R⁰ K` where
+`K` is a fraction field of `R`. The norm is defined by
+`FractionalIdeal.absNorm I = Ideal.absNorm I.num / |Algebra.norm ℤ I.den|` where `I.num` is a
+ideal of `R` and `I.den` an element of `R⁰` such that `I.den • I = I.num`.
+
+## Main definitions and results
+
+ * `FractionalIdeal.absNorm`: the norm as a zero preserving morphism with values in `ℚ`.
+ * `FractionalIdea.absNorm_eq'`: the value of the norm does not depend on the choice of
+   `I.num` and `I.den`.
+ * `FractionalIdeal.abs_det_basis_change`: the norm is given by the determinant
+    of the basis change matrix.
+ * `FractionalIdeal.absNorm_span_singleton`: the norm of a principal fractional ideal is the
+   norm of its generator
+-/
+
+
+variable {R : Type*} [CommRing R]
+
+namespace FractionalIdeal
+
+open scoped Pointwise nonZeroDivisors
+
+variable [IsDomain R] [Infinite R] [IsDedekindDomain R] [Module.Free ℤ R] [Module.Finite ℤ R]
```

Comment:
I know you didn't introduce this, but I really think we should have `Nontrivial R` (or maybe this is even automatic since it is a dedekind domain) instead of `Infinite R`: since it is free this is enough.

I don't think Lean is able to find `Infinite R` by itself, but you can add the instance if you don't want to remove `Infinite R` everywhere. Or maybe just add a TODO if this is too much work for one PR.

Reply:
Yeah, the hypothesis `Infinite R` is used only at one place in `Ideal.absNorm` and should indeed be replaced by `Nontrivial R` and the result that `Module.Free S R` + `Infinite S` + `Nontrivial R` implies `Infinite R`. I'll open a PR with this change to merge before this one.

Reply:
#9684